src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs

```diff
@@ -1074,10 +1093,81 @@ private bool BelongsToGroup()
             return WorkspaceViewModel.Model.Annotations.ContainsModel(this.annotationModel);
         }
 
+        /// <summary>
+        /// Determines whether this group displays warning or error icons in its header.
+        /// </summary>
+        private void UpdateErrorAndWarningIconVisibility()
+        {
+            // No icons are displayed when the group is expanded / not collapsed.
+            if (IsExpanded)
+            {
+                GroupState = ElementState.Active;
+                return;
+            }
+
+            // Fetching all nodes inside the group.
+            List<NodeModel> nodeModels = Nodes
+                .OfType<NodeModel>()
+                .ToList();
+
+            // Subscribing to changes in the nodes' state. If a node's state 
+            // changes (e.g. to warning) we'll want to reflect this.
+            foreach (NodeModel nodeModel in nodeModels)
+            {
+                nodeModel.PropertyChanged += NodeModelOnPropertyChanged;
+            }
+
+            // Adding to this nodes from any nested groups.
+            nodeModels.AddRange
+            (
+                Nodes.OfType<AnnotationModel>()
+                    .Select(x => x.Nodes)
+                    .SelectMany(x => x)
+                    .OfType<NodeModel>()
+            );
+
+            // If any nodes (even in nested groups) are in error state we display an error icon.
+            if (nodeModels.Any(x => x.State == ElementState.Error))
+            {
+                GroupState = ElementState.Error;
+                return;
+            }
+
+            // If any nodes (even in nested groups) are in warning state we display a warning icon.
+            if (nodeModels.Any(x => x.State == ElementState.Warning))
+            {
+                GroupState = ElementState.Warning;
+                return;
+            }
+
+            GroupState = ElementState.Active;
+        }
+
+        /// <summary>
+        /// If a node's State changes (e.g. to, or from, a Warning state) we'll want the UI to reflect
+        /// this and show/hide a warning icon in the group header to reflect this.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void NodeModelOnPropertyChanged(object sender, PropertyChangedEventArgs e)
+        {
+            switch (e.PropertyName)
+            {
+                case nameof(NodeModel.State):
+                    UpdateErrorAndWarningIconVisibility();
+                    break;
+            }
+        }
+
+
         public override void Dispose()
         {
             annotationModel.PropertyChanged -= model_PropertyChanged;
             annotationModel.RemovedFromGroup -= OnModelRemovedFromGroup;
+            foreach (NodeModel nodeModel in Nodes.OfType<NodeModel>())
```

Comment:
is this also done when a node is removed from a group?

Reply:
Seems like the `AnnotationViewModel` method `HandleNodesCollectionChanges` isn't picking up the `removedFromGroup` and `addedToGroup` nodes correctly. This would be the best place to subscribe/unsubscribe nodes from the event.

Reply:
This is taken care of with existing `model_PropertyChanged` on `AnnotationModel` now.